File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonSourceParser.java
package com.nytimes.android.external.store.middleware;


import com.google.gson.Gson;
import com.nytimes.android.external.store.base.Parser;

import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.nio.charset.Charset;

import javax.inject.Inject;

import io.reactivex.annotations.NonNull;
import okio.BufferedSource;

import static com.nytimes.android.external.cache.Preconditions.checkNotNull;


/**
 * Parser to be used when going from a BufferedSource to any Parsed Type
 * example usage:
 * ParsingStoreBuilder.<BufferedSource, BookResults>builder()
 * .fetcher(fetcher)
 * .persister(SourcePersisterFactory.create(getApplicationContext().getCacheDir()))
 * .parser(GsonParserFactory.createSourceParser(new Gson(),BookResult.class)
 * .open();
 */


public class GsonSourceParser<Parsed> implements Parser<BufferedSource, Parsed> {

    private final Gson gson;
    private final Type type;

    @Inject
    public GsonSourceParser(Gson gson, Type type) {
        checkNotNull(gson, "Gson can't be null");
        checkNotNull(type, "Type can't be null");
        this.gson = gson;
        this.type = type;
    }

    @Override
    @SuppressWarnings({"PMD.EmptyCatchBlock"})
    public Parsed apply(@NonNull BufferedSource bufferedSource) {
        try (InputStreamReader reader = new InputStreamReader(bufferedSource.inputStream(), Charset.forName("UTF-8"))) {
            return gson.fromJson(reader, type);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: middleware/src/main/java/com/nytimes/android/external/store3/middleware/GsonSourceParser.kt
package com.nytimes.android.external.store3.middleware


import com.google.gson.Gson
import com.nytimes.android.external.store3.base.Parser
import com.nytimes.android.external.store3.util.ParserException
import okio.BufferedSource
import java.io.IOException
import java.io.InputStreamReader
import java.lang.reflect.Type
import java.nio.charset.Charset
import javax.inject.Inject


/**
 * Parser to be used when going from a BufferedSource to any Parsed Type
 * example usage:
 * ParsingStoreBuilder.<BufferedSource></BufferedSource>, BookResults>builder()
 * .fetcher(fetcher)
 * .persister(SourcePersisterFactory.create(getApplicationContext().getCacheDir()))
 * .parser(GsonParserFactory.createSourceParser(new Gson(),BookResult.class)
 * .open();
 */


class GsonSourceParser<Parsed> @Inject
constructor(private val gson: Gson, private val type: Type) : Parser<BufferedSource, Parsed> {

    @Throws(ParserException::class)
    override suspend fun apply(raw: BufferedSource): Parsed {
        try {
            InputStreamReader(raw.inputStream(), Charset.forName("UTF-8")).use { reader -> return gson.fromJson(reader, type) }
        } catch (e: IOException) {
            throw ParserException(e?.message ?: "", e)
        }

    }
}
